### PR TITLE
Promise(Like)-compatible RSVP type definitions.

### PR DESCRIPTION
**Pull request by chriskrycho**
## Reference

- [RSVP implementation]
    - [all exports]
- [Previous type defs]

[RSVP implementation]: https://github.com/tildeio/rsvp.js/tree/v4.6.1/lib/rsvp
[all exports]: https://github.com/tildeio/js/blob/v4.6.1/lib/js#L73-L94
[Previous type defs]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/rsvp/index.d.ts

## Exports

Derived from the list of [all exports]:

- [x] `asap`
- [x] `cast` – deprecated alias of `resolve()`
- [x] `Promise`
    - [x] constructor
    - [x] `.all()`
    - [x] `.catch()`
    - [x] `.finally()`
    - [x] `.race()`
    - [x] `.reject()`
    - [x] `.cast()` – deprecated alias of `resolve()`
    - [x] `.resolve()`
    - [x] `.then()`
- [x] `EventTarget`
- [x] `all`
- [x] `allSettled`
- [x] `race`
- [x] `hash`
- [x] `hashSettled`
- [x] `rethrow`
- [x] `defer`
- [ ] `denodeify`
- [ ] `configure`
- [x] `on`
- [x] `off`
- [x] `resolve`
- [x] `reject`
- [x] `map`
- [x] `async`
- [x] `filter`

The `RSVP` namespace itself should simply be the [default export](https://github.com/tildeio/rsvp.js/blob/c1be42ece7e9854e4744894c6b2da3609417aa3a/lib/rsvp.js#L50-L71), reexporting those items as an object.

### CODE REVIEW

**Comment by chriskrycho** [comment]
**Note:** until we *also* rewrite the Ember type defs – *again* – to take only one type parameter for RSVP Promises, this will fail everything. UGH.

**Comment by chriskrycho** [comment]
I fixed the remaining type defs issue mentioned above with e0243b8; at this point it *should* just be a matter of doing the (annoying, but mundane) work of adding the remaining definitions with tests. That's mostly just a lot of boilerplate to get it right—I noticed some quirks with the smart but ultimately not-sufficiently-polymorphic approach @theroncross and I took with e.g. `RSVP.hash` last time.

File: types/rsvp/rsvp-tests.ts [hunk 1]
```diff
@@ -118,7 +118,29 @@ function testHash() {
 }
 
 function testHashSettled() {
-    // TODO: add test
+    function isFulfilled<T>(state: RSVP.PromiseState<T>): state is RSVP.Resolved<T> {
```
Comment (chriskrycho) [inline types/rsvp/rsvp-tests.ts]: Oh, this is nice; I'm going to steal it and use it over in the `allSettled` in short order.

**Comment by chriskrycho** [comment]
I'm pretty sure `denodeify` is going to kill me. I'm seriously thinking of doing what Bluebird does and not even *trying* to do the variants which handle turning extra Node arguments into either hashes or extra items in a tuple.

**Comment by dwickern** [comment]
Let's start with some tests

**Comment by chriskrycho** [comment]
Oh, I have *lots* of tests. I’ll ship just those for now.

**Comment by chriskrycho** [comment]
At this point, I'm pretty okay with where things stand. Unfortunately, I can't see a good way to accomplish two things with `denodeify` and this:

- correctly *requiring* (not merely *allowing*) the `this` context to be bound if you do `RSVP.denodeify(foo.bar)` – the `foo` context gets lost.
- adding additional arguments to the Node functions being denodeified, i.e. everything *before* the callback. Bluebird gets around this by supporting more of those but only one callback value. We could go either way, I think; but RSVP seems in its options biased somewhat toward supporting more callback values (since it has not one but two different ways to handle having more than one of those).

I think I'm good with merging this, doing a little local testing, and then upstreaming it with the relevant qualifications and caveats to RSVP proper.

**Comment by dwickern** [comment]
I was able to get `this` context working with a pair overloads like these (one for `void`, another for some generic `this`)
```ts
function denodeify<T1, T2, A, K1 extends string, K2 extends string>(
    nodeFunc: (this: void, arg1: A, callback: (err: any, data1: T1, data2: T2) => void) => void,
    options: [K1, K2]
): ( arg1: A) => RSVP.Promise<{ [K in K1]: T1 } & { [K in K2]: T2 }>;

function denodeify<This, T1, T2, A, K1 extends string, K2 extends string>(
    nodeFunc: (this: This, arg1: A, callback: (err: any, data1: T1, data2: T2) => void) => void,
    options: [K1, K2]
): (this: This, arg1: A) => RSVP.Promise<{ [K in K1]: T1 } & { [K in K2]: T2 }>;
```

**Comment by chriskrycho** [comment]
Ah, awesome. I see what the problem was with mine; I was trying to do it with a *generic* `This` bound as `void`, and I can see how that wouldn't work. It wouldn't inherently know which overload to resolve.p since they're the same type arity.

**Comment by chriskrycho** [comment]
Follow-up: if you can get a PR where it's working, I'll of course merge it, @dwickern, but my local experiments are *not* getting everything to play nicely across the board—possibly because of a combination of the type arity + declaration ordering issues TS has?

**Pull request merged**